File: api/src/main/java/org/gtri/util/iteratee/api/Planner.java
/*
    Copyright 2012 Georgia Tech Research Institute

    Author: lance.gatlin@gtri.gatech.edu

    This file is part of org.gtri.util.iteratee library.

    org.gtri.util.iteratee library is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    org.gtri.util.iteratee library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with org.gtri.util.iteratee library. If not, see <http://www.gnu.org/licenses/>.

*/

package org.gtri.util.iteratee.api;

/**
 * An interface that it used to plan streaming operations on Producer, 
 * Consumer and Translator objects.
 * 
 * @author lance.gatlin@gmail.com
 */
public interface Planner {
  
  /**
   * Get the factory that was used to create this planner. 
   * @return the factory that was used to create this planner. 
   */
  Factory factory();
  
  /**
   * Create a plan to stream items from a producer to a consumer.
   * 
   * @param <A> the item type
   * @param producer 
   * @param consumer
   * @return a plan that can be run to obtain results
   */
  <A> Consumer.Plan<A,A> connect(Producer<A> producer, Consumer<A> consumer);
  
  /**
   * Create a plan to stream items from a producer to a consumer by translating
   * the items using the given translator.
   * 
   * @param <A> the produced type
   * @param <B> the consumed type
   * @param producer 
   * @param translator
   * @param consumer
   * @return a plan that can be run to obtain results
   */
  <A,B> Consumer.Plan<A,B> connect(Producer<A> producer, Translator<A,B> translator, Consumer<B> consumer);
  
  /**
   * Create a plan to stream items from a producer to an iteratee.
   * 
   * @param <A> the item type
   * @param producer 
   * @param iteratee
   * @return a plan that can be run to obtain results
   */
  <A,S> Iteratee.Plan<A,A,S> connect(Producer<A> producer, Iteratee<A,S> iteratee);
  
  /**
   * Create a plan to stream items from a producer to an iteratee by translating
   * the items using the given translator.
   * 
   * @param <A> the producer output type
   * @param <B> the iteratee input type
   * @param producer 
   * @param translator
   * @param iteratee
   * @return a plan that can be run to obtain results
   */
  <A,B,S> Iteratee.Plan<A,B,S> connect(Producer<A> producer, Translator<A,B> translator, Iteratee<B,S> iteratee);
  
  /**
   * Create a plan to stream items from a producer to a builder.
   * 
   * @param <A> the item type
   * @param producer 
   * @param builder
   * @return a plan that can be run to obtain results
   */
  <A,V> Builder.Plan<A,A,V> connect(Producer<A> producer, Builder<A,V> builder);

  /**
   * Create a plan to stream items from a producer to a builder by translating
   * the items using the given translator.
   * 
   * @param <A> the produced type
   * @param <B> the type consumed by the builder
   * @param producer 
   * @param translator
   * @param builder
   * @return a plan that can be run to obtain results
   */
  <A,B,V> Builder.Plan<A,B,V> connect(Producer<A> producer, Translator<A,B> translator, Builder<B,V> builder);
  
  /**
   * Compose two translators.
   * 
   * @param <A> the input type of the first translator
   * @param <B> the output type of the first translator and input type of the 
   * second
   * @param <C> the output type of the second translator 
   * @param first first translator
   * @param second second translator
   * @return a translator composed of the two translators
   */
  <A,B,C> Translator<A,C> compose(Translator<A,B> first, Translator<B,C> second);
  
  /**
   * Compose three translators.
   * 
   * @param <A> the input type of the first translator
   * @param <B> the output type of the first translator and input type of the 
   * second
   * @param <C> the output type of the second translator and input type of the 
   * third
   * @param <D> the output type of the third translator 
   * @param first first translator
   * @param second second translator
   * @param third third translator
   * @return a translator composed of the three translators
   */
  <A,B,C,D> Translator<A,D> compose(
    Translator<A,B> first, 
    Translator<B,C> second,
    Translator<C,D> third
  );
  
  /**
   * Compose four translators.
   * 
   * @param <A> the input type of the first translator
   * @param <B> the output type of the first translator and input type of the 
   * second
   * @param <C> the output type of the second translator and input type of the 
   * third
   * @param <D> the output type of the third translator and input type of the 
   * fourth
   * @param <E> the output type of the fourth translator 
   * @param first first translator
   * @param second second translator
   * @param third third translator
   * @param fourth fourth translator
   * @return a translator composed of the four translators
   */
  <A,B,C,D,E> Translator<A,E> compose(
    Translator<A,B> first, 
    Translator<B,C> second,
    Translator<C,D> third,
    Translator<D,E> fourth
  );
  
  /**
   * Compose five translators.
   * 
   * @param <A> the input type of the first translator
   * @param <B> the output type of the first translator and input type of the 
   * second
   * @param <C> the output type of the second translator and input type of the 
   * third
   * @param <D> the output type of the third translator and input type of the 
   * fourth
   * @param <D> the output type of the fourth translator and input type of the 
   * fifth
   * @param <E> the output type of the fifth translator 
   * @param first first translator
   * @param second second translator
   * @param third third translator
   * @param fourth fourth translator
   * @param fifth fifth translator
   * @return a translator composed of the five translators
   */
  <A,B,C,D,E,F> Translator<A,F> compose(
    Translator<A,B> first, 
    Translator<B,C> second,
    Translator<C,D> third,
    Translator<D,E> fourth,
    Translator<E,F> fifth
  );
  
}

File: impl/src/main/scala/org/gtri/util/iteratee/impl/Planner.scala
/*
    Copyright 2012 Georgia Tech Research Institute

    Author: lance.gatlin@gtri.gatech.edu

    This file is part of org.gtri.util.iteratee library.

    org.gtri.util.iteratee library is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    org.gtri.util.iteratee library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with org.gtri.util.iteratee library. If not, see <http://www.gnu.org/licenses/>.

*/

package org.gtri.util.iteratee.impl

import scala.collection.immutable.Traversable
import org.gtri.util.iteratee.api
import api._
import translate._
import util.Looper


/**
* Created with IntelliJ IDEA.
* User: Lance
* Date: 11/11/12
* Time: 1:11 PM
* To change this template use File | Settings | File Templates.
*/
class Planner(val factory : api.Factory) extends api.Planner {
  val self = this
  def connect[A](p: Producer[A], c : Consumer[A]) =
    new Consumer.Plan[A,A] {

      def producer = p

      def consumer = c

      def run = {
        // TODO: find a way to do this without identity translatee
        val looper = new Looper[A,A,Unit,ProducerState[A],ConsumerState[A]](p.initialState, new IdentityTranslator[A]().initialState, c.initialState)
        val lastL = looper.run(factory.issueHandlingCode())
        val lastP = lastL.enumerator
        val lastC = lastL.machine
        // Feed EOI to lastI to force it to complete
        val eoiResult = lastC.endOfInput()

        new Consumer.Result[A,A] {

          def state = eoiResult.output

          def status = eoiResult.next.status

          def issues = eoiResult.issues ++ looper.issues

          def overflow = eoiResult.overflow ++ looper.overflow

          def producer = factory.createProducer(lastP)

          def consumer = factory.createConsumer(lastC)
        }
      }
    }

  def connect[A, B](p: Producer[A], t: Translator[A, B], c: Consumer[B]) =
    new Consumer.Plan[A,B] {

      def producer = p

      def consumer = c

      def run = {
        val looper = new Looper[A,B,Unit,ProducerState[A],ConsumerState[B]](p.initialState, t.initialState, c.initialState)
        val lastL = looper.run(factory.issueHandlingCode())
        val lastP  = lastL.enumerator
        val lastC = lastL.machine
        // Feed EOI to lastI to force it to complete
        val eoiResult = lastC.endOfInput()

        new Consumer.Result[A,B] {

          def state = eoiResult.output

          def status = eoiResult.next.status

          def issues = eoiResult.issues

          def overflow = eoiResult.overflow

          def producer = factory.createProducer(lastP)

          def consumer = factory.createConsumer(lastC)
        }
      }
    }

  def connect[A,S](p: Producer[A], i : Iteratee[A,S]) =
    new Iteratee.Plan[A,A,S] {

      def producer = p

      def iteratee = i

      def run = {
        // TODO: find a way to do this without identity translatee
        val looper = new Looper[A,A,S,ProducerState[A], IterateeState[A,S]](p.initialState, new IdentityTranslator[A]().initialState, i.initialState)
        val lastL = looper.run(factory.issueHandlingCode())
        val lastP = lastL.enumerator
        val lastI = lastL.machine
        // Feed EOI to lastI to force it to complete
        val eoiResult = lastI.endOfInput()

        new Iteratee.Result[A,A,S] {

          def loopState = eoiResult.output

          def status = eoiResult.next.status

          def issues = eoiResult.issues ++ looper.issues

          def overflow = eoiResult.overflow ++ looper.overflow

          def producer = factory.createProducer(lastP)

          def iteratee = factory.createIteratee(lastI)
        }
      }
    }

  def connect[A, B, S](p: Producer[A], t: Translator[A, B], i: Iteratee[B,S]) =
    new Iteratee.Plan[A,B, S] {

      def producer = p

      def iteratee = i

      def run = {
        val looper = new Looper[A,B,S,ProducerState[A],IterateeState[B,S]](p.initialState, t.initialState, i.initialState)
        val lastL = looper.run(factory.issueHandlingCode())
        val lastP  = lastL.enumerator
        val lastI = lastL.machine
        // Feed EOI to lastI to force it to complete
        val eoiResult = lastI.endOfInput()

        new Iteratee.Result[A,B,S] {

          def loopState = eoiResult.output

          def status = eoiResult.next.status

          def issues = eoiResult.issues

          def overflow = eoiResult.overflow

          def producer = factory.createProducer(lastP)

          def iteratee = factory.createIteratee(lastI)
        }
      }
    }

  def connect[A, V](p: Producer[A], b: Builder[A, V]) =
    new Builder.Plan[A,A,V] {

      def producer = p

      def builder = b

      def run = {
        // TODO: find a way to do this without identity translatee
        val looper = new Looper[A,A,Option[V],ProducerState[A], BuilderState[A,V]](p.initialState, new IdentityTranslator[A]().initialState, b.initialState)
        val lastL = looper.run(factory.issueHandlingCode())
        val lastP = lastL.enumerator
        val lastB = lastL.machine
        // Feed EOI to lastI to force it to complete
        val eoiResult = lastB.endOfInput()

        new Builder.Result[A,A,V] {
          def value = eoiResult.output

          def status = eoiResult.next.status

          def issues = eoiResult.issues

          def overflow = eoiResult.overflow

          def producer = factory.createProducer(lastP)

          def builder = factory.createBuilder(lastB)
        }
      }
    }


  def connect[A, B, V](p: Producer[A], t: Translator[A, B], b: Builder[B, V]) =
    new Builder.Plan[A,B,V] {

      def producer = p

      def builder = b

      def run = {
        val looper = new Looper[A,B,Option[V],ProducerState[A],BuilderState[B,V]](p.initialState, t.initialState, b.initialState)
        val lastL = looper.run(factory.issueHandlingCode())
        val lastP  = lastL.enumerator
        val lastB = lastL.machine
        // Feed EOI to lastI to force it to complete
        val eoiResult = lastB.endOfInput()

        new Builder.Result[A,B,V] {

          def value = eoiResult.output

          def status = eoiResult.next.status

          def issues = eoiResult.issues

          def overflow = eoiResult.overflow

          def producer = factory.createProducer(lastP)

          def builder = factory.createBuilder(lastB)
        }
      }
    }


  def compose[A, B, C](t1: Translator[A, B], t2: Translator[B, C]) = TranslatorTuple2(t1,t2)

  def compose[A, B, C, D](t1: Translator[A, B], t2: Translator[B, C], t3: Translator[C, D]) = TranslatorTuple3(t1,t2,t3)

  def compose[A, B, C, D, E](t1: Translator[A, B], t2: Translator[B, C], t3: Translator[C, D], t4: Translator[D, E]) = TranslatorTuple4(t1,t2,t3,t4)

  def compose[A, B, C, D, E, F](t1: Translator[A, B], t2: Translator[B, C], t3: Translator[C, D], t4: Translator[D, E], t5: Translator[E, F]) = TranslatorTuple5(t1,t2,t3,t4,t5)
}
